8310242: Clarify the name parameter to Class::forName

mlchung opened the pull request:
This PR clarifies the spec of the 3-arg Class::forName  regarding the format of the name for an array type which is of the form: one or more of "[" + binary name of the element type + ";'.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8310244](https://bugs.openjdk.org/browse/JDK-8310244) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8310242](https://bugs.openjdk.org/browse/JDK-8310242): Clarify the name parameter to Class::forName (**Bug** - P4)
 * [JDK-8310244](https://bugs.openjdk.org/browse/JDK-8310244): Clarify the name parameter to Class::forName (**CSR**)


### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [004f35d5](https://git.openjdk.org/jdk/pull/14528/files/004f35d591e1e61ce725a0dea2e2775d1e204234)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - Author)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [852bd774](https://git.openjdk.org/jdk/pull/14528/files/852bd7740fb7a83e3b15b128a6d319efe4aae17b)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14528/head:pull/14528` \
`$ git checkout pull/14528`

Update a local copy of the PR: \
`$ git checkout pull/14528` \
`$ git pull https://git.openjdk.org/jdk.git pull/14528/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14528`

View PR using the GUI difftool: \
`$ git pr show -t 14528`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14528.diff">https://git.openjdk.org/jdk/pull/14528.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14528#issuecomment-1595241362)

RogerRiggs commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -438,8 +438,10 @@ private static Class<?> forName(String className, Class<?> caller)
      *
-     * <p> If {@code name} denotes an array class, the component type of
-     * the array class is loaded but not initialized.
+     * <p> If {@code name} denotes an array class, i.e. the name consists of
+     * one or more {@code '['} followed by the binary name of the element type
+     * and {@code ';'}, the element type of the array class is loaded but
+     * not initialized.
```
Do arrays of primitive types need special treatment?  For example, "[I"

mlchung replied:
Thanks for catching it.  Updated.

RogerRiggs commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -438,10 +438,15 @@ private static Class<?> forName(String className, Class<?> caller)
-     * and {@code ';'}, the element type of the array class is loaded but
-     * not initialized.
+     * <p> If {@code name} denotes an array class whose element type
+     * is a class or interface, i.e. the name consists of one or more {@code '['}
+     * followed by the binary name of the element type and {@code ';'}
+     * the element type of the array class is loaded but not initialized.
+     *
+     * <p> To find an array class of a primitive type, the name string
+     * consists of one or more {@code '['} followed by the field type descriptor
+     * of the primitive type (JVMS {@jvms 4.3.2 }) e.g. {@code "[I"} to
```
Can we avoid mixing "i.e." and "e.g"?
There are more "For example" occurrences than either "i.e." or "e.g." 
Your call.

mlchung replied:
ok.  is the revised version better?

RogerRiggs reviewed (approved):
Looks good, thanks

liach commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -438,8 +438,15 @@ private static Class<?> forName(String className, Class<?> caller)
-     * the array class is loaded but not initialized.
+     * <p> If {@code name} denotes an array class whose element type
+     * is a class or interface, the name consists of one or more {@code '['}
+     * followed by the binary name of the element type and {@code ';'},
```
There's an L between [ and component binary name, like [Ljava.lang.Object;

mlchung replied:
Updated.

AlanBateman reviewed (approved):
This changes look okay but they do make me wonder if we should add a note to the 2-arg Class.forName to make it clear that it does not allow class names for array types, the class name has to be a binary name.

jensli commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -392,9 +392,10 @@ static String typeVarBounds(TypeVariable<?> typeVar) {
-     * @param      className   the fully qualified name of the desired class.
-     * @return     the {@code Class} object for the class with the
-     *             specified name.
+     * @param     className the <a href="#binary-name">binary name</a> of the class
```
The new text is probably more correct, but it is less easy to understand. I think "fully qualified name" of a class is a well-known term, but "binary name" is rather unknown.

I suggest changing the text to mention both terms. For example like this:

```suggestion
     * @param     className  the <a href="#binary-name">binary name</a>) of the class, for example its fully qualified name
```

liach replied:
Binary name is defined in JLS. Fully qualified name is ambiguous as it may refer to the internal name or a descriptor string.

jensli commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -392,9 +392,10 @@ static String typeVarBounds(TypeVariable<?> typeVar) {
-     * @param      className   the fully qualified name of the desired class.
-     * @return     the {@code Class} object for the class with the
-     *             specified name.
+     * @param     className the <a href="#binary-name">binary name</a> of the class
```
Where do these links lead? I don't find any section with this ID in the `Class` Javadoc. Should they lead to `ClassLoader`? That Javadoc has such a section.

dholmes-ora replied:
This seems a specification change as the binary-name and FQN are not always the same. In particular for a member class `M` in class `C` the binary name is `C$M`, whereas the FQN is `C.M`.

AlanBateman replied:
> Binary name is a long-standing behavior. It's a spec bug.

Yes, I view this PR as just specifying long standing behavior. The name used for array class may be surprising but I don't think it can be changed.

jensli commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -423,11 +424,10 @@ private static Class<?> forName(String className, Class<?> caller)
-     * locate and load the class or interface.  The specified class
-     * loader is used to load the class or interface.  If the parameter
-     * {@code loader} is null, the class is loaded through the bootstrap
+     * Given the {@linkplain ##binary-name binary name} for a class or interface,
```
Is double hash a mistake?

ExE-Boss replied:
That’s a new **Javadoc** feature available since **JDK 20**: [JDK‑8200337] ([GH‑10395])

[JDK‑8200337]: https://bugs.openjdk.org/browse/JDK-8200337 "[JDK‑8200337] Generalize see and link tags for user‑defined anchors"
[GH‑10395]: https://github.com/openjdk/jdk/pull/10395

mlchung commented:
I added an API note in the 2-arg forName that it does not support loading of array types.  Also fixed the links.   A new test is added to show the valid and invalid names.

mlchung commented:
I improved the spec of `forName` to make it clear of the name for nested types and array types.   It now refers to the table in `getName` about the name of an array class.   

@dholmes-ora does that clarify and address the concerns you had?

dholmes-ora commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -423,45 +424,52 @@ private static Class<?> forName(String className, Class<?> caller)
-     * the array class is loaded but not initialized.
-     *
-     * <p> For example, in an instance method the expression:
+     * <p> To obtain {@code Class} object associated with an array class,
+     * the name consists of one or more {@code '['} representing the depth
+     * of the array class, followed by the element type as encoded in
```
Suggestion
```
+     * <p> To obtain the {@code Class} object associated with an array class,
+     * the name consists of one or more {@code '['} representing the depth
+     * of the array nesting, followed by the element type as encoded in
```

dholmes-ora commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -423,45 +424,52 @@ private static Class<?> forName(String className, Class<?> caller)
-     * Class.forName("Foo")
+     * Class<?> threadClass = Class.forName("java.lang.Thread", false, currentLoader);
+     * Class<?> stringArrayClass = Class.forName("[Ljava.lang.String;", false, currentLoader);
+     * Class<?> intArrayClass = Class.forName("[[[I", false, currentLoader);
```
The variable name doesn't match the actual class loaded. It might be clearer to add a trailing comment `// Class of int[][][]` ?

dholmes-ora commented on test/jdk/java/lang/Class/forName/ForNameNames.java:
```diff
@@ -0,0 +1,84 @@
+
+    @ParameterizedTest
+    @MethodSource("testCases")
+    void testForName(String cn, Class<?> expected) throws ClassNotFoundException {
```
Should this also test that `c.getName().equals(cn)`?

dholmes-ora reviewed (changes requested):
A few small suggestions, otherwise looks good.

Thanks.

liach commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -423,45 +424,52 @@ private static Class<?> forName(String className, Class<?> caller)
     /**
      * Returns the {@code Class} object associated with the class or
      * interface with the given string name, using the given class loader.
```
Should we update the summary to `... associated with the class or interface or array with the given string name...`

mlchung replied:
I want to keep it as is.  The first paragraph talks about a class or interface and the other paragraphs talk about array classes, primitive types and hidden classes.

mlchung commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -423,45 +424,52 @@ private static Class<?> forName(String className, Class<?> caller)
+     * <p> This method cannot be used to obtain any of the {@code Class} objects
+     * representing primitive types or void, hidden classes or interfaces,
+     * or array classes whose element type is a hidden class or interface.
+     * If {@code name} denotes a primitive type or void, for example {@code I},
```
One main point for this example to show is that it will find a class in unnamed package named `I`.

mlchung commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -423,45 +424,52 @@ private static Class<?> forName(String className, Class<?> caller)
-     * {@snippet lang="java" :
-     * Class.forName("Foo", true, this.getClass().getClassLoader())
-     * }
+     * <p> A call to {@code forName("[L}<i>N</i>{@code ;")} causes the element type
```
I considered that.   The spec and the example already describe multi-dimensional array.  I think one can infer from this single-dimensional array example.

dholmes-ora commented on src/java.base/share/classes/java/lang/Class.java:
```diff
@@ -441,14 +441,14 @@ private static Class<?> forName(String className, Class<?> caller)
      *
      * <p> To obtain {@code Class} object associated with an array class,
```
To obtain _the_ Class object ...

AlanBateman commented on test/jdk/java/lang/Class/forName/ForNameNames.java:
```diff
@@ -0,0 +1,92 @@
+        return Stream.of(
+                Arguments.of("java.lang.String", String.class),
+                Arguments.of("[Ljava.lang.String;", String[].class),
+                Arguments.of("ForNameNames$Inner", Inner.class),
```
For completes, you could add "[LForNameNames$Inner;" to this list of test cases.

mlchung commented:
I attached the specdiff to the [CSR](https://bugs.openjdk.org/browse/JDK-8310244) - please review.

mlchung commented:
/integrate
